### PR TITLE
update groups when creating a new project(AST-64011)

### PR DESCRIPTION
**Pull request by miryamfoiferCX**
By submitting a PR to this repository, you agree to the terms within the [Checkmarx Code of Conduct](../docs/code_of_conduct.md). Please see the [contributing guidelines](../docs/contributing.md) for how to create and submit a high-quality PR for this repo.

### Description

> update groups when creating a new project, in the first request of creation

### References

> https://checkmarx.atlassian.net/browse/AST-64011

### Testing

> refactor 2 unit-tests that don't get now the project ID due to the fact that we broke the flow before the create
>add a new unit-test to check a case of creating successfully a new project with 2 existing groups

### Checklist

- [x] I have added documentation for new/changed functionality in this PR (if applicable).
- [x] I have updated the CLI help for new/changed functionality in this PR (if applicable).
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

File: internal/services/groups.go [hunk 1]
```diff
@@ -8,6 +8,20 @@ import (
...
+func GetGroupMap(groupsWrapper wrappers.GroupsWrapper, projectGroups string, projModel *wrappers.Project, projModelResp *wrappers.ProjectResponseModel,
+	featureFlagsWrapper wrappers.FeatureFlagsWrapper) ([]*wrappers.Group, error) {
+	groupsMap, groupErr := CreateGroupsMap(projectGroups, groupsWrapper)
+	if groupErr != nil {
+		return nil, errors.Errorf("%s: %v", failedUpdatingProj, groupErr)
+	}
+	projModel.Groups = getGroupsForRequest(groupsMap, featureFlagsWrapper)
+	if projModelResp != nil {
+		groups := append(getGroupsForRequest(groupsMap, featureFlagsWrapper), projModelResp.Groups...)
+		projModel.Groups = groups
```
Comment (AlvoBen) [inline internal/services/groups.go:20]: I think it's better that this function will only return the group map and will not change the projectModel object.

File: internal/services/projects.go [hunk 2]
```diff
@@ -87,6 +87,12 @@ func createProject(
 	var projModel = wrappers.Project{}
 	projModel.Name = projectName
 	projModel.ApplicationIds = applicationID
+	if projectGroups != "" {
+		_, groupErr := GetGroupMap(groupsWrapper, projectGroups, &projModel, nil, featureFlagsWrapper)
```
Comment (AlvoBen) [inline internal/services/projects.go:91]: related to previous comment:
groupMap, groupErr = GetGroupMap(groupsWrapper, projectGroups, nil, featureFlagsWrapper)
projModel.Groups = groupMap

File: internal/services/projects.go [hunk 3]
```diff
@@ -215,25 +221,36 @@ func updateProject(
...
+	accessManagementWrapper wrappers.AccessManagementWrapper, projModelResp *wrappers.ProjectResponseModel,
+	projectGroups string, projectID string, projectName string, featureFlagsWrapper wrappers.FeatureFlagsWrapper) error {
+	return UpsertProjectGroups(groupsWrapper, projModel, projectsWrapper, accessManagementWrapper, projModelResp, projectGroups, projectID, projectName, featureFlagsWrapper)
+}
+
+func UpsertProjectGroupsByUpdateFlow(groupsWrapper wrappers.GroupsWrapper, projModel *wrappers.Project, projectsWrapper wrappers.ProjectsWrapper,
 	accessManagementWrapper wrappers.AccessManagementWrapper, projModelResp *wrappers.ProjectResponseModel,
 	projectGroups string, projectID string, projectName string, featureFlagsWrapper wrappers.FeatureFlagsWrapper) error {
-	groupsMap, groupErr := CreateGroupsMap(projectGroups, groupsWrapper)
+	_, groupErr := GetGroupMap(groupsWrapper, projectGroups, projModel, projModelResp, featureFlagsWrapper)
```
Comment (AlvoBen) [inline internal/services/projects.go:241]: is it not duplication? why do you call here GetGroupMaps if you also have this call in UpsertProjectGroups?

File: internal/services/projects_test.go [hunk 4]
```diff
@@ -114,7 +123,7 @@ func Test_createProject(t *testing.T) {
 			accessManagementWrapper: &mock.AccessManagementMockWrapper{},
 			projectGroups:           "grp1,grp2",
 			featureFlagsWrapper:     &mock.FeatureFlagsMockWrapper{},
-		}, want: "ID-new-project-name", wantErr: true},
+		}, want: "", wantErr: true},
```
Comment (AlvoBen) [inline internal/services/projects_test.go:126]: Why did you change this test case?

File: internal/services/projects_test.go [hunk 5]
```diff
@@ -132,7 +141,7 @@ func Test_createProject(t *testing.T) {
 			accessManagementWrapper: &mock.AccessManagementMockWrapper{},
 			projectGroups:           "fake-group-error",
 			featureFlagsWrapper:     &mock.FeatureFlagsMockWrapper{},
-		}, want: "ID-new-project-name", wantErr: true},
+		}, want: "", wantErr: true},
```
Comment (AlvoBen) [inline internal/services/projects_test.go:144]: also here

File: internal/services/projects.go [hunk 6]
```diff
@@ -215,25 +221,36 @@ func updateProject(
...
+		err = UpsertProjectGroupsByUpdateFlow(groupsWrapper, &projModel, projectsWrapper, accessManagementWrapper, projModelResp, projectGroups, projectID, projectName, featureFlagsWrapper)
 		if err != nil {
 			return projectID, err
 		}
 	}
 	return projectID, nil
 }
 
-func UpsertProjectGroups(groupsWrapper wrappers.GroupsWrapper, projModel *wrappers.Project, projectsWrapper wrappers.ProjectsWrapper,
+func UpsertProjectGroupsByCreateFlow(groupsWrapper wrappers.GroupsWrapper, projModel *wrappers.Project, projectsWrapper wrappers.ProjectsWrapper,
```
Comment (AlvoBen) [inline internal/services/projects.go:232]: related to next comment, what is the difference between the update and the create flow?

**Comment by AlvoBen** [comment]
Please also include Jira ticket reference

File: internal/services/projects.go (hunk 6 above)
Comment (miryamfoiferCX) [reply internal/services/projects.go:232]: in the update we should call first to GetGroupMap, and than continue calling to UpsertProjectGroups
but in create flow we did it at the creation part at the begin of the flow, so no need to call this logic twice

File: internal/services/projects_test.go (hunk 4 above)
Comment (miryamfoiferCX) [reply internal/services/projects_test.go:126]: before the fix, we fetch the groups after creating the project, so also in a case of error we returned projectID
now we're fetching groups before the creation so in case of failure we still don't have projectID and in this case we are getting an empty string

File: internal/services/projects_test.go (hunk 5 above)
Comment (miryamfoiferCX) [reply internal/services/projects_test.go:144]: same answer as above

File: internal/services/projects.go (hunk 6 above)
Comment (AlvoBen) [reply internal/services/projects.go:232]: But your first call in UpsertProjectGroups is to GetGroupMap func and both of the upsert flows (update/create) are calling UpsertProjectGroups function

File: internal/services/projects.go (hunk 3 above)
Comment (miryamfoiferCX) [reply internal/services/projects.go:241]: my bad
forgot to remove it from UpsertProjectGroups

File: internal/services/projects.go (hunk 2 above)
Comment (miryamfoiferCX) [reply internal/services/projects.go:91]: fixed in a similar way

File: internal/services/groups.go (hunk 1 above)
Comment (miryamfoiferCX) [reply internal/services/groups.go:20]: fixed

File: internal/services/projects.go (hunk 6 above)
Comment (miryamfoiferCX) [reply internal/services/projects.go]: that what I wrote there - my bad forgot to remove this call
fixed

File: internal/services/projects.go [hunk 7]
```diff
@@ -215,27 +225,34 @@ func updateProject(
...
+		err = UpsertProjectGroupsByUpdateFlow(groupsWrapper, &projModel, projectsWrapper, accessManagementWrapper, projModelResp, projectGroups, projectID, projectName, featureFlagsWrapper)
 		if err != nil {
 			return projectID, err
 		}
 	}
 	return projectID, nil
 }
 
-func UpsertProjectGroups(groupsWrapper wrappers.GroupsWrapper, projModel *wrappers.Project, projectsWrapper wrappers.ProjectsWrapper,
+func UpsertProjectGroupsByCreateFlow(projModel *wrappers.Project, projectsWrapper wrappers.ProjectsWrapper,
```
Comment (AlvoBen) [inline internal/services/projects.go:236]: Now this function do nothing but passing the parameters to UpsertProjectGroups func - delete it and use the UpsertProjectGroups directly

**Review by AlvoBen — APPROVED** [review]
